server/src/main/java/org/elasticsearch/search/query/QueryPhaseTimeoutException.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.query;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.search.SearchShardTarget;
+
+import java.io.IOException;
+
+/**
+ * Specific instance of QueryPhaseExecutionException that indicates that a search timeout occurred.
+ * Always returns http status 408 (request timeout)
+ */
+public class QueryPhaseTimeoutException extends QueryPhaseExecutionException {
+    public QueryPhaseTimeoutException(SearchShardTarget shardTarget, String msg, Throwable cause) {
+        super(shardTarget, "Query Failed [" + msg + "]", cause);
+    }
+
+    public QueryPhaseTimeoutException(StreamInput in) throws IOException {
+        super(in);
+    }
+
+    public QueryPhaseTimeoutException(SearchShardTarget shardTarget, String msg) {
+        super(shardTarget, msg);
+    }
+
+    @Override
+    public RestStatus status() {
+        return RestStatus.REQUEST_TIMEOUT;
```

Comment:
I don't think it's correct to use `408 Request Timeout` in this context. See for instance [RFC 2616 section 10.4.9](https://datatracker.ietf.org/doc/html/rfc2616#section-10.4.9):

```
408 Request Timeout

   The client did not produce a request within the time that the server
   was prepared to wait. The client MAY repeat the request without
   modifications at any later time.
```

That's not what's happening here, we (the server) are not waiting for the client to produce a request, the client is waiting for us to respond.

(I know that we already use 408 in a few places in ES but these are bugs, and hard to fix because of BwC concerns; let's not add to the problem here at least)

Reply:
do you have another 4xx code in mind that would be a better fit for a request timeout? The point here is that the user decided to not accept partial results and fail the entire request in case a single shard times out.

Reply:
Typically a server-side timeout would be a 5xx response, even if it timed out because the client asked the server to do an unreasonable amount of work. I believe `403 Forbidden` and `404 Not Found` could be correct in this situation too:

```
403 Forbidden

   The server understood the request, but is refusing to fulfill it.
   Authorization will not help and the request SHOULD NOT be repeated.
   If the request method was not HEAD and the server wishes to make
   public why the request has not been fulfilled, it SHOULD describe the
   reason for the refusal in the entity.  If the server does not wish to
   make this information available to the client, the status code 404
   (Not Found) can be used instead.

404 Not Found

   The server has not found anything matching the Request-URI. No
   indication is given of whether the condition is temporary or
   permanent. The 410 (Gone) status code SHOULD be used if the server
   knows, through some internally configurable mechanism, that an old
   resource is permanently unavailable and has no forwarding address.
   This status code is commonly used when the server does not wish to
   reveal exactly why the request has been refused, or when no other
   response is applicable.
```

Reply:
Also I believe we could get away with `202 Accepted`:

```
202 Accepted

   The request has been accepted for processing, but the processing has
   not been completed.  The request might or might not eventually be
   acted upon, as it might be disallowed when processing actually takes
   place. There is no facility for re-sending a status code from an
   asynchronous operation such as this.

   The 202 response is intentionally non-committal. Its purpose is to
   allow a server to accept a request for some other process (perhaps a
   batch-oriented process that is only run once per day) without
   requiring that the user agent's connection to the server persist
   until the process is completed. The entity returned with this
   response SHOULD include an indication of the request's current status
   and either a pointer to a status monitor or some estimate of when the
   user can expect the request to be fulfilled.
```

Frankly I'd stick with a 5xx code tho. HTTP status codes aren't designed for this kind of finesse, we need some other mechanism to distinguish this case from other 5xx situations.

Reply:
I don't like any of these options :)  I think using 403 and 404 is going to be confusing because they are generally used in many places to indicate something very different from what is meant in this case. 202 looks a bit exotic and hard to understand. 500 is weird, because here it's the user that specifically asked to get an error instead of partial results. I do get the concerns on 408, the description does not match what happens. 

Is 504 - gateway timeout perhaps an option?

Reply:
```
504 Gateway Timeout

   The server, while acting as a gateway or proxy, did not receive a
   timely response from the upstream server specified by the URI (e.g.
   HTTP, FTP, LDAP) or some other auxiliary server (e.g. DNS) it needed
   to access in attempting to complete the request.
```

I think that's valid in this situation, yes. The coordinating node is in some sense acting as a gateway, translating the client's request into requests to the underlying shards. I will say that I don't totally like the idea, we get support cases about 504 responses received because of unreasonable timeouts in users' proxies and it's super useful to be able to say "Elasticsearch never itself returns a 504, you need to look at your infra" without having to dig deeper, but that's not a watertight argument against making this change if you're really certain it's needed.

Reply:
Thanks for the feedback, I agree it's not perfect, but I think 504 in place of 500 is more indicative of this situation and we can benefit from the distinction. We do have plans to also track metrics on returned status codes directly in ES, which should help the situation when it comes to identifying whether a certain status code has been returned by ES or by something else.

Reply:
The latest push changes the RestStatus from 408 to 504.